internal/controllers/configuration/kongupstreampolicy_utils.go

```diff
@@ -0,0 +1,217 @@
+func (r *KongUpstreamPolicyReconciler) enforceKongUpstreamPolicyStatus(ctx context.Context, oldPolicy *kongv1beta1.KongUpstreamPolicy) (bool, error) {
+	// get all the services that reference this UpstreamPolicy
+	services := &corev1.ServiceList{}
+	err := r.List(ctx, services,
+		client.InNamespace(oldPolicy.Namespace),
+		client.MatchingFields{
+			upstreamPolicyIndexKey: oldPolicy.Name,
+		},
+	)
+	if err != nil {
+		return false, err
+	}
+
+	// build the desired KongUpstreamPolicy status
+	servicesStatus, err := r.buildServicesStatus(ctx, k8stypes.NamespacedName{
+		Namespace: oldPolicy.Namespace,
+		Name:      oldPolicy.Name,
+	}, services.Items)
+	if err != nil {
+		return false, err
+	}
+
+	newPolicyStatus := gatewayapi.Policystatus{}
+	if len(servicesStatus) > 0 {
+		newPolicyStatus.Ancestors = make([]gatewayapi.PolicyAncestorStatus, 0, len(servicesStatus))
+	}
+	for _, ss := range servicesStatus {
+		newPolicyStatus.Ancestors = append(newPolicyStatus.Ancestors,
+			gatewayapi.PolicyAncestorStatus{
+				AncestorRef: gatewayapi.ParentReference{
+					Group:     lo.ToPtr(gatewayapi.Group("core")),
+					Kind:      lo.ToPtr(gatewayapi.Kind("Service")),
+					Namespace: lo.ToPtr(gatewayapi.Namespace(ss.service.Namespace)),
+					Name:      gatewayapi.ObjectName(ss.service.Name),
+				},
+				ControllerName: gatewaycontroller.GetControllerName(),
+				Conditions: []metav1.Condition{
+					ss.acceptedCondition,
+				},
+			},
+		)
+	}
+	if isStatusUpdated := isPolicyStatusUpdated(oldPolicy.Status, newPolicyStatus); !isStatusUpdated {
+		newPolicy := oldPolicy.DeepCopy()
+		newPolicy.Status = newPolicyStatus
+		return true, r.Client.Status().Patch(ctx, newPolicy, client.MergeFrom(oldPolicy))
+	}
+	return false, nil
+}
+
+// buildServicesStatus creates a list of services with their conditions associated.
+func (r *KongUpstreamPolicyReconciler) buildServicesStatus(ctx context.Context, upstreamPolicyNN k8stypes.NamespacedName, services []corev1.Service) ([]serviceStatus, error) {
+	// sort the services by creationTimestamp
+	sort.Slice(services, func(i, j int) bool {
+		return services[i].CreationTimestamp.Before(&services[j].CreationTimestamp)
+	})
+
+	// prepare a service mapping to be used in subsequent operations
+	mappedServices := make(map[string]serviceStatus)
+	for i, service := range services {
+		acceptedCondition := metav1.Condition{
+			Type:               string(gatewayapi.PolicyConditionAccepted),
+			Status:             metav1.ConditionTrue,
+			Reason:             string(gatewayapi.PolicyReasonAccepted),
+			LastTransitionTime: metav1.Now(),
+		}
+		if i < maxNAncestors {
+			mappedServices[buildServiceReference(service.Namespace, service.Name)] = serviceStatus{
+				service:           service,
+				acceptedCondition: acceptedCondition,
+			}
+		} else {
+			r.Log.Info("kongUpstreamPolicy %s/%s status has already %d ancestors, cannot set service %s/%s as an ancestor in the status",
+				upstreamPolicyNN.Namespace,
```

Comment:
The `logr.Log` does not support `Infof` to accept format strings. Please use `Info(fmt.Sprintf(...))` or 
```
Info(
".....",
"namepace", upstreamPolicyNN,
...
)
```
.